proposals/0110-exponential-fee-for-write-lock-accounts.md

```diff
@@ -0,0 +1,154 @@
+  - Accounts are associated with a *compute unit pricer*, and the *runtime*
+  maintains an LRU cache of actively contentious accounts' public keys and
+  their *compute unit pricers*.
+  - Alternatively, each account can have its *compute unit pricer* stored
+  onchain, which would require modifying accounts.
+- Compute Unit Pricer:
+  - Tracks an account's EMA *compute-unit utilization*, updated after the
+  current bank is frozen.
+  - Provides the current *cost rate* when queried.
+- EMA of Compute-Unit Utilization:
+  - Uses 8 slots for EMA calculation.
+  - EMA Alpha, representing the degree of weighting decrease, is calculated as
+  `alpha = 2 / (N+1)`.
+- Pricing Algorithm:
+  - Adjusts write-lock *cost rate* based on an account's EMA *compute-unit
+  utilization*.
+  - For each block, if an account's EMA *compute-unit utilization* is more than
+  target utilization, its write-lock *cost rate* increases by X%. If it's
+  below target, the *cost rate* decreases by X%.
+  - For V0:
+    - target utilization is `25%` of Max limit;
+    - Initial write-lock cost rate is `1000 micro-lamport/CU`;
+    - cost change rate set to 1%;
+- Calculate *Write Lock Fee*:
+  - Fee required to write-lock an account is calculated by multiplying the
+  write-lock *cost rate* by the transaction's requested CU.
+
+### Detailed Design
+
+- Initialization and Inheritance:
+  - Bank initializes an empty account_write_lock_fee_cache, an Cache of
+  {account_pubkey, compute-unit-pricer}.
+  - Child banks inherit the parent's cache.
+- Transaction Fee Calculation:
+  - Calculate write-lock fee for each account a transaction needs to write,
+  summing up to be its *write lock fee*. This, along with signature fee and
+  priority fee, constitutes the total fee for the transaction.
+  - Leader checks fee payer's balance before scheduling the transaction.
+- Cost Tracking:
+  - Cost_tracker tracks CUs for the current block and each write-locked account
+  as-is;
+  - Ensuring cost tracking is enabled at the replay stage.
+- End of Block Processing:
+  - Identify write-locked accounts with *compute-unit utilization* > target
+  utilization. Add/update bank's account_write_lock_fee_cache.
+  - For those accounts are in cache, eg were saturated, but not being
+  write-locked in this block, their current block *compute-unit utilization*
+  shall be considered as zero; then use it to update Cache.
+  - Evict cheapest account before add new "hot" accounts into LRU cach, if cache
+  is in full capacity;
+  - Account will be evicted from cache as soon as its *cost rate* drop back to 0
+  - Cache has capacity set to 2* worst case eviction per block to prevent
+  cache attack.
+  - For v0, cache capacity set to 2048, as:
+    - Max number of tansactions with account 6M CU = 48M/6M = 8;
+    - Max number of accounts per tx: 128;
+    - worst case per block: 128 * 8 = 1024;
+    - 2 times worst case: 2048;
+- Fee Handling:
+  - Collected write-lock fees are 100% burnt.
```

Comment:
Why aren't read locks considered here?

Reply:
`Other Considerarion` section acknowledges read-lock contention and possibility of applying same EMA mechanism. Prefer doing that after seeing success on write-lock first.